File: self_library/src/main/java/com/wjf/self_library/view/codingUI/BaseUi.java
package com.wjf.self_library.view.codingUI;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;

import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;


/**
 * @author Wangjf2-DESKTOP
 */
public abstract class BaseUi<T extends ViewDataBinding> extends ConstraintLayout {
    protected T view;
    protected Context context;

    protected BaseUi(Context context, AttributeSet attrs) {
        super(context, attrs);
        this.context = context;
        View root = View.inflate(context, setLayout(), this);
        LayoutInflater inflater=LayoutInflater.from(context);
        view = DataBindingUtil.inflate(inflater,setLayout(),this,false);
        getTypeArray(context.obtainStyledAttributes(attrs, setStyleable()));
        setView();
    }

    /**
     * 设置布局
     *
     * @return 布局资源id
     */
    protected abstract int setLayout();

    /**
     * 设置样式
     *
     * @return 返回样式资源id
     */
    protected abstract int[] setStyleable();

    /**
     * 获取样式类型
     *
     * @param typedArray 样式数组
     */
    protected abstract void getTypeArray(TypedArray typedArray);

    /**
     * 设置View
     */
    protected abstract void setView();
}

File: corekit/src/main/java/org/jxxy/debug/corekit/common/BaseUi.kt
package org.jxxy.debug.corekit.common

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.viewbinding.ViewBinding

/**
 * @author : Wangjf
 * @date : 2021/1/19
 */
abstract class BaseUi<T : ViewBinding> protected constructor(
    context: Context,
    attrs: AttributeSet?
) : ConstraintLayout(context, attrs) {
    constructor(context: Context) : this(context, null)

    protected lateinit var view: T
        private set

    /**
     * 设置布局
     *
     * @return 布局资源id
     */
    protected abstract fun bindLayout(inflater: LayoutInflater, parent: ViewGroup): T

    /**
     * 设置样式
     *
     * @return 返回样式资源id
     */
    protected abstract fun setStyleable(): IntArray

    /**
     * 获取样式类型
     *
     * @param typedArray 样式数组
     */
    protected abstract fun getTypeArray(typedArray: TypedArray)

    /**
     * 设置View
     */
    protected abstract fun setView()

    private fun initWidget(attrs: AttributeSet?) {
        view = bindLayout(LayoutInflater.from(context), this)
        getTypeArray(context.obtainStyledAttributes(attrs, setStyleable()))
        setView()
    }

    init {
        initWidget(attrs)
    }
}
